modules/standard/Sort.chpl

```diff
@@ -24,45 +24,103 @@
+
+  sort(Array);
+
+  // This will output: 1, 1, 2, 3, 3, 4, 5, 5, 5, 6, 9
+  writeln(Array);
+
+
+The sort function can also accept a region argument to sort a subset of an
+array. This is offered as an optimization over using an array slice which may
+have performance overhead.
+
+
+.. code-block:: chapel
+
+  var Array = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5];
+
+  // Sort only the elemets in the range 1..5
+  // Same as sort(Array[1..5]);
+  sort(Array, region=1..5);
+
+  // This will output: 3, 1, 1, 4, 5, 9, 2, 6, 5, 3, 5
+  writeln(Array);
+
+
+The sort function can also be called on a list, be stable or unstable, and
+accept a custom comparator.
+See the :proc:`sort(x: list)<Sort.sort>` function for details.
 
 .. _comparators:
 
 Comparators
 -----------
 
 Comparators allow sorting data by a mechanism other than the
-default comparison operations between array elements. To use a comparator,
-define a record or a class with an appropriate method and then pass
-an instance of it to the sort function. Examples are shown below.
+default comparison operations between array elements.
+
+The :proc:`sort` function can accept a comparator argument, which defines how
+the data is sorted. If no comparator is passed, the default comparator is
+used.
 
-Comparators need to include at least one of the following methods:
+Reverse sorting is handled by the :record:`ReverseComparator`.
+See :ref:`Reverse Comparator<reverse-comparator>` for details.
 
- * ``key(a)``  -- see `The .key method`_
- * ``compare(a, b)``  -- see `The .compare method`_
- * ``keyPart(a, i)`` -- see `The .keyPart method`_
 
-See the section below for discussion of each of these methods.
+To use a custom comparator, define a record or a class which implements the
+appropriate sorting interface.
 
-A comparator can contain both ``compare`` and ``keyPart`` methods. In that
-event, the sort algorithm will use whichever is appropriate for the algorithm
-and expect that they have consistent results.
+Comparators need to implement one, and only one, of the following interfaces
+as well as at least one of their associated methods:
+
+  * :interface:`keyComparator` -- see `The keyComparator interface`_
+  * :interface:`relativeComparator` -- see `The relativeComparator interface`_
+  * :interface:`keyPartComparator` -- see `The keyPartComparator interface`_
+
+See the section below for discussion of each of these interfaces and methods.
+
+*Future:*
+
+  Provide a unified ``sortComparator`` interface, which can represent an
+  exclusive or (XOR) of the three interfaces above.
+
+
+The keyComparator interface
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``keyComparator`` interface is used to sort data by a key value. Records
+implementing this interface must define a ``key`` method.
 
 It is an error for a comparator to contain a ``key`` method as well as one of
-the other methods.
+the other methods that are part of the ``relativeComparator`` or
+``keyPartComparator`` interfaces.
+
```

Comment:
This isn't strictly true. With interfaces, a comparator can have both a key and compare method, but it can only implement one interface.
```chapel
record cmp: keyComparator {
  // key method for keyComparator
  proc key ...
  // arbitrary method, users can write whatever they want in this record
  proc compare ...
```

Reply:
https://github.com/jabraham17/chapel/blob/64b8a6d12cbd564be0a504231b9b5f051454be1b/modules/standard/Sort.chpl#L443-L462

This seems to suggest that we can't have a `.key` and a `.compare` method in a record that implements `keyComparator`

Reply:
I would call that an artifcact of implementation. With the deprecation logic removed, you will be able to have a `.compare` in a keyComparator, it just won't be a special method

Reply:
I see what you mean. But today it is still an error then, and it is intentionally so. Therefore the docs should mention it?

Reply:
I can phrase it like this

```
Today, it is an error for a comparator to contain a ``key`` method as well as one of
the other methods that are part of the ``relativeComparator`` or
``keyPartComparator`` interfaces. This restriction might be lifted in future releases.
```

Reply:
Also we have future work for `sortComparator` planned which is an XOR of all three interfaces:
```chapel
record cmp: sortComparator {
  // Is this one valid 
  proc key ...
  // Or is this one valid
  proc compare ...
```
Although this is definitely not relevant for this PR

Reply:
I like that phrasing